src/Storageless/Http/SessionHijackingMitigationMiddleware.php

```diff
@@ -0,0 +1,106 @@
+ * and is licensed under the MIT license.
+ */
+
+declare(strict_types=1);
+
+namespace PSR7Sessions\Storageless\Http;
+
+use Psr\Http\Message\ResponseInterface as Response;
+use Psr\Http\Message\ServerRequestInterface as Request;
+use Psr\Http\Server\MiddlewareInterface;
+use Psr\Http\Server\RequestHandlerInterface;
+use PSR7Sessions\Storageless\Session\SessionInterface;
+use RuntimeException;
+
+use function array_key_exists;
+use function assert;
+use function base64_encode;
+use function is_string;
+use function sodium_crypto_generichash;
+use function sprintf;
+
+final class SessionHijackingMitigationMiddleware implements MiddlewareInterface
+{
+    public const SERVER_PARAM_REMOTE_ADDR = 'REMOTE_ADDR';
+    public const SERVER_PARAM_USER_AGENT  = 'HTTP_USER_AGENT';
+
+    public const SESSION_KEY = 'fp';
+
+    /** @param literal-string $sessionAttribute */
+    public function __construct(
+        private readonly string $sessionAttribute = SessionMiddleware::SESSION_ATTRIBUTE,
+    ) {
+    }
+
+    public function process(Request $request, RequestHandlerInterface $handler): Response
+    {
+        $session = $request->getAttribute($this->sessionAttribute);
+        assert($session instanceof SessionInterface);
+        $fingerprint = $this->getFingerprint($request);
+
+        if (! $session->isEmpty()) {
+            if ($fingerprint !== $session->get(self::SESSION_KEY)) {
+                $session->clear();
+            }
+        }
+
+        $response = $handler->handle($request);
+
+        if (! $session->isEmpty()) {
+            $session->set(self::SESSION_KEY, $fingerprint);
+        }
+
+        return $response;
+    }
+
+    /** @return non-empty-string */
+    private function getFingerprint(Request $request): string
+    {
+        $serverParams = $request->getServerParams();
+        if (
+            ! array_key_exists(self::SERVER_PARAM_REMOTE_ADDR, $serverParams)
+            || ! is_string($serverParams[self::SERVER_PARAM_REMOTE_ADDR])
+            || $serverParams[self::SERVER_PARAM_REMOTE_ADDR] === ''
+        ) {
+            throw new RuntimeException(sprintf(
+                'The request lacks a valid %s parameter',
+                self::SERVER_PARAM_REMOTE_ADDR,
+            ));
+        }
+
+        if (
+            ! array_key_exists(self::SERVER_PARAM_USER_AGENT, $serverParams)
+            || ! is_string($serverParams[self::SERVER_PARAM_USER_AGENT])
+            || $serverParams[self::SERVER_PARAM_USER_AGENT] === ''
+        ) {
+            throw new RuntimeException(sprintf(
+                'The request lacks a valid %s parameter',
+                self::SERVER_PARAM_USER_AGENT,
+            ));
+        }
+
+        $fingerprint = base64_encode(sodium_crypto_generichash(sprintf(
+            '%s#%s',
+            $serverParams[self::SERVER_PARAM_REMOTE_ADDR],
+            $serverParams[self::SERVER_PARAM_USER_AGENT],
+        )));
+        assert($fingerprint !== '');
```

Comment:
Why not use the token claims (and validate them)?

This populates data inside the token, rather than using those, which were designed precisely for this kind of use-case.

Reply:
That's indeed what I hoped for at the beginning, but:

1. I'd like to keep it as a separate plugable middleware
2. If so, I' don't have access to the `lcobucci\jwt` object, but only to the `\PSR7Sessions\Storageless\Session\SessionInterface`

Reply:
Could we box/unbox the JWT token here? Mostly seeing duplicate crypto operations, which is more of a weakness than strength, IMO :thinking: 

The alternative is baking this logic inside the pre-existing middleware, which indeed increases complexity a lot, and requires some conditionals for enabling/disabling.

I need to think about this at architecture level: operating with a middleware makes indeed sense, but this operates after the session has already been accepted, which IMO is already a problem if any other middleware operate on that session too.

Reply:
Mmm, a separate plugable middleware is easier to understand, develop and maintain, but indeed not strong enough.

I can refactor the change to add (yet) another constructor parameter to `SessionMiddleware`.

P.S.: for `psr7-sessions/storageless:v9` we should really need to move to a `SessionMiddlewareConfiguration` object, `SessionMiddleware::__construct` is getting out of control.